### PR TITLE
Issue/1372

### PR DESCRIPTION
**Pull request by alexvins**

### CODE REVIEW

File: server/CGameHandler.cpp [hunk 1]
```diff
@@ -5764,13 +5765,52 @@ void CGameHandler::duelFinished()
...
+		}
+	};
+
+	//1. Find removed stacks.
+	for(const auto & slotInfo : army->stacks)
+	{
+		const SlotID slot = slotInfo.first;
+		const CStackInstance * instance = slotInfo.second;
+
+		if(nullptr != instance)//just in case
```
Comment (DjWarmonger) [inline server/CGameHandler.cpp]: Isn't
if (!instance) simpler in this case?

Comment (alexvins) [inline server/CGameHandler.cpp]: I don`t like pointer->bool conversion at all.

**Comment by ArseniyShestakov** [comment]
Ok current implementation need improvements:
 - Stack that used for "resurrection" into Demons not appear on Battle Casualties list while it's appear in H3.
 - No check what corpses can be used for "ressurection". In H3 there status bar text e.g in case of elementals "Can only summon Demons from the corpses of living creatures". I suppose limitations of this spell should be about the same as with normal resurrection, but personally I once checked that it's not work for elementals, undead.
- It's shouldn't be castable on not suitable stacks. E.g in H3 there is status bar message "The dead %s do not have sufficient life-force to summon Demons". With current implementation if stack not suitable you can still attempt to cast it and if it's won't work Pit Lords don't lose spell cast, but lose turn.

BTW as always there issue with summoned elementals so I finally posted related bug:
http://bugs.vcmi.eu/view.php?id=2304

**Comment by ArseniyShestakov** [comment]
Also it's looks like everything else is working fine.

**Comment by ArseniyShestakov** [comment]
One more tricky moment that current code handle wrong way which results into stack with 0 creatures:
* Get hero with base Resurrection spell (this one only resurrect creatures till end of battle)
* Get him army of pit lords and some easy to kill stack sufficient for summon.
* Go into battle and let this stack die, resurrect it into demons.
* Now let demons die and resurrect them using spell.
* Finish battle and you'll end up with zero creatures stack.

Though I no idea if it's bug within Resurrection or this summon code.

**Comment by alexvins** [comment]
Displaying removed stacks in battle result requires deep refactoring of battle code. This is different task.

**Comment by ArseniyShestakov** [comment]
I expected there may be problems with battle casualties check as I think there is more places where it's not work properly. Though at least allowed creatures check need to be there before merging that. Corpses limitation needed, but it's not as game-breaking.

File: client/battle/CBattleInterface.cpp [hunk 2]
```diff
@@ -2163,7 +2163,14 @@ void CBattleInterface::handleHex(BattleHex myNumber, int eventType)
...
 				if (shere && ourStack && !shere->alive())
-					legalAction = true;
+				{
+					if(!(shere->hasBonusOfType(Bonus::UNDEAD) 
+						|| shere->hasBonusOfType(Bonus::NON_LIVING) 
+						|| vstd::contains(shere->state, EBattleStackState::SUMMONED)
+						|| vstd::contains(shere->state, EBattleStackState::CLONED)
+						))
+						legalAction = true;
+				}
```
Comment (ArseniyShestakov) [inline client/battle/CBattleInterface.cpp]: BTW is there any particular reason why RISE_DEMONS need it's own code at all?
I just see that Archangel resurrection somehow handled by ANY_CREATURE and it's working fine.

Comment (ArseniyShestakov) [inline client/battle/CBattleInterface.cpp]: Oops. Got it. It's not really implemented as actual spell...

**Comment by DjWarmonger** [comment]
Is this ready to go? Would like to merge before 0.98e.

**Comment by ArseniyShestakov** [comment]
It's still possible to summon daemons from war machines which clearly has to be fixed before merge.

Though if dead stack health insufficient you no longer lose turn so this can be easily fixed some next time once this implemented as actual spell.

**Pull request merged**